### PR TITLE
Write-through enabled ram Db

### PR DESCRIPTION
**Pull request by DrizztDoUrden**
This is revived PR https://github.com/AMDComputeLibraries/MLOpen/pull/1968

### CODE REVIEW

**Comment by atamazov** [comment]
Description updated.

**Comment by atamazov** [comment]
Oops... wait

File: src/include/miopen/sqlite_db.hpp [hunk 1]
```diff
@@ -278,38 +282,50 @@ class SQLiteBase
         return AllFound;
     }
     template <typename... U>
-    inline auto FindRecord(U&... args)
+    inline boost::optional<DbRecord> FindRecord(U&... args)
     {
+        if(DisableDbFileIO)
```
Comment (atamazov) [inline src/include/miopen/sqlite_db.hpp:287]: This change disables both layers of SQLite-based system perf-db, which is `MultiFileDb<SQLitePerfDb, SQLitePerfDb, true>`, because `SQLitePerfDb` inherits from this class.

We shall keep the SQLite-based system perf-db in working state (or it is disabled in the baseline `develop`???)

Comment (atamazov) [inline src/include/miopen/sqlite_db.hpp:287]: This also affects the SQLite-based kernel cache `KernDb : public SQLiteBase<KernDb>`, but it shouldn't

**Comment by atamazov** [comment]
Description updated.

**Comment by DrizztDoUrden** [comment]
@atamazov resolved the comments

**Comment by DrizztDoUrden** [comment]
Although, I am not 100% sure I should disable user KkernDb.

File: src/ramdb.cpp [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    static std::mutex mutex;
+    static const std::lock_guard<std::mutex> lock{mutex};
+
+    static bool saved      = false;
+    static auto saved_path = path;
+    static RamDb instance{path, warn_if_unreadable};
+
+    if(saved)
+    {
+        if(saved_path != path)
```
Comment (JehandadKhan) [inline src/ramdb.cpp:109]: This means MIOpen would not support a system with two different GPUs. Can't we just return the instance for that GPU instead ?

File: src/db.cpp [hunk 3]
```diff
@@ -62,14 +57,17 @@ PlainTextDb::PlainTextDb(const std::string& filename_,
                          const std::size_t /*num_cu*/)
     : PlainTextDb(filename_, is_system)
 {
+    if(is_system)
```
[the remark this answers is not in the document]
Comment (DrizztDoUrden) [reply src/db.cpp:60]: No, but at the moment uncached PlainTextDb shouldn't be used directly anywhere. All the system plain text Dbs should be ReadOnlyRamDb and user - RamDb.

[the remark this answers is not in the document]
Comment (JehandadKhan) [reply src/db.cpp:60]: @atamazov frequently uses the plain text db as perf db by using the compile time args. This would break that capability.

File: src/ramdb.cpp [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    if(!file)
+    {
+        MIOPEN_LOG_E("Cannot update database modification time: " + time_file_path);
+        return;
+    }
+
+    file << time.count();
+}
+
+#define MIOPEN_VALIDATE_LOCK(lock)                       \
```
[the remark this answers is not in the document]
Comment (DrizztDoUrden) [reply src/ramdb.cpp:80]: It is defined in db.cpp, and I would prefer not to move it to db.hpp, which is indirectly included in almost the whole library due to the fact we use dbs as templates everywhere instead of just inheriting with an interface class or at least type erasure.

File: src/ramdb.cpp [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    } while(false)
+
+static std::chrono::seconds GetLockTimeout() { return std::chrono::seconds{60}; }
+
+using exclusive_lock = std::unique_lock<LockFile>;
+
+RamDb::RamDb(std::string path, bool is_system) : PlainTextDb(path, is_system)
+{
+    if(is_system)
+        MIOPEN_THROW("PlainTextDb class is not supported as system database. Use the ReadOnlyRamDb "
```
[the remark this answers is not in the document]
Comment (DrizztDoUrden) [reply src/ramdb.cpp:94]: Hm, this throw is not required at all as the PlainTextDb will throw anyway.

File: src/ramdb.cpp (hunk 2 above)
Comment (DrizztDoUrden) [reply src/ramdb.cpp:109]: We currently don't support running on two different types of GPU. We had an ancient discussion with Artem about this. I can change it to a map if we have a plan to support multiple different GPUs in any foreseeable future.

File: src/ramdb.cpp [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+                     "class instead.");
+}
+
+RamDb& RamDb::GetCached(const std::string& path, bool warn_if_unreadable)
+{
+    static std::mutex mutex;
+    static const std::lock_guard<std::mutex> lock{mutex};
```
[the remark this answers is not in the document]
Comment (DrizztDoUrden) [reply src/ramdb.cpp:101]: Yeah, sorry, ancient copy-paste from read-only class. Will fix RN.

File: src/include/miopen/ramdb.hpp [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+    bool ValidateUnsafe();
+    void Prefetch();
+
+#if MIOPEN_DB_CACHE_WRITE_THROUGH
+    void UpdateCacheEntryUnsafe(const DbRecord& record);
+#endif
+};
+
+/// \todo This is modified copy of code from db.hpp. Make a proper fix.
```
[the remark this answers is not in the document]
Comment (DrizztDoUrden) [reply src/include/miopen/ramdb.hpp:144]: DbTimer is outsidish for RamDb. I want to measure the prefetch here.

File: src/db.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (DrizztDoUrden) [reply src/db.cpp:60]: No, not really, I have updated the text PerfDb to cached as there is no benefit to not cache:
```
using PerformanceDb = DbTimer<MultiFileDb<ReadonlyRamDb, RamDb, true>>;
```

**Comment by pfultz2** [comment]
Where is the performance comparison of this compared to sqlite?

**Comment by DrizztDoUrden** [comment]
At the moment there is none of current state. There is one in AMDComputeLibraries/MLOpen/pull/1968 dated as 2019.10.12.

**Comment by atamazov** [comment]
> Although, I am not 100% sure if I should disable user KernDb.

See https://github.com/AMDComputeLibraries/MLOpen/issues/2219. It introduces the notions of user and system binary caches.

Basically, we need a feature that should should switch off the user binary cache. I am not sure if this already implemented (under control of `MIOPEN_DISABLE_USERDB`), but this PR should not change this aspect of behavior.

The system binary cache is read-only and intended to hold pre-compiled binaries.

**Comment by atamazov** [comment]
@pfultz2 
> Where is the performance comparison of this compared to sqlite?

We do not have sqlite-based find-db. We can compare only cached and non-cached text based find databases. The comparison is available at https://github.com/AMDComputeLibraries/MLOpen/pull/1968.

File: src/include/miopen/ramdb.hpp [hunk 8]
```diff
@@ -0,0 +1,211 @@
+/*******************************************************************************
+*
+* MIT License
+*
+* Copyright (c) 2019 Advanced Micro Devices, Inc.
```
[the remark this answers is not in the document]
Comment (atamazov) [reply src/include/miopen/ramdb.hpp]: In review for years ;)

File: src/ramdb.cpp (hunk 2 above)
Comment (atamazov) [reply src/ramdb.cpp:106]: Yes, please see https://github.com/AMDComputeLibraries/MLOpen/pull/1968#discussion_r319216292.

Comment (atamazov) [reply src/ramdb.cpp:106]: We have a dedicated issues about that, https://github.com/AMDComputeLibraries/MLOpen/issues/1312

**Comment by pfultz2** [comment]
> We do not have sqlite-based find-db.

We should probably implement that first and then see how the performance compares.

**Comment by DrizztDoUrden** [comment]
Should I implement sqlite find-db?

**Comment by atamazov** [comment]
> > We do not have sqlite-based find-db.
> 
> We should probably implement that first and then see how the performance compares.

Why ___first___? Testing shows obvious performance gain vs current develop. And we have it unmerged for months. How long we are going to delay this?

**Comment by pfultz2** [comment]
> Why first?

I dont see why we need to bring in this complexity when sqlite already does this and is better tested.

**Comment by atamazov** [comment]
> > Why first?
> 
> I dont see why we need to bring in this complexity when sqlite already does this and is better tested.

If you do not see why we need this: We need this for better performance. Find-db lookup time directly adds to Wall time in Immediate mode and to Aux Wall time in Hybrid Find.

What Sqlite already does? Delivers performance? Let's test and compare. I have ___nothing against it___ if it can provide comparable performance.

Why do you think sqlite-based solution is better tested?

Why do you think sqlite already does this? find-db logic is different from perf-db. And there is no sqlite based implementation yet available.

Evaluation of complexity is a value judgement. You think that Sqlite is simpler, but I do think that Sqlite based implementation is more complex than text based one.

File: src/include/miopen/sqlite_db.hpp (hunk 1 above)
Comment (atamazov) [reply src/include/miopen/sqlite_db.hpp]: Hmm... Now I am not sure, see https://github.com/ROCmSoftwarePlatform/MIOpen/pull/280#issuecomment-644931011

File: src/ramdb.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (atamazov) [reply src/ramdb.cpp]: [Notice] It's ok, but I didn't expected that PlainTextDb cannot be used as a system db anymore.

**Comment by pfultz2** [comment]
> What Sqlite already does? Delivers performance?

Most likely performance. Sqlite doesn't load the entire database into memory(unlike the text db), and it will cache the chunks that have been read.

It also provides correctness, since its caching works even when modifying the db.

> Why do you think sqlite-based solution is better tested?

* Four independently developed test harnesses
* 100% branch test coverage in an as-deployed configuration
* Millions and millions of test cases
* Out-of-memory tests
* I/O error tests
* Crash and power loss tests
* Fuzz tests
* Boundary value tests
* Disabled optimization tests
* Regression tests
* Malformed database tests
* Extensive use of assert() and run-time checks
* Valgrind analysis
* Undefined behavior checks
* Checklists 

This testing goes way beyond the testing we do for our db files. We have had MT issues that shown up and they are not easy to debug, so if we can utilize something that has been well tested, it will be better and save us debugging/development time.

We can still tweak sqlite to get better performance which will be better than inventing our own read/write caching that is not well tested.

**Comment by atamazov** [comment]
> > What Sqlite already does? Delivers performance?
> 
> Most likely performance. Sqlite doesn't load the entire database into memory(unlike the text db), and it will cache the chunks that have been read.

I am afraid that our use case (find-db) wouldn't  allow SQLite to show its performance supremacy. Let's see when we have SQLite-based find-db.

> > Why do you think sqlite-based solution is better tested?
> 
> - Four independently developed test harnesses
>  ...

Well, all this is true. But this describes how well SQLite itseif has been tested. But I am more interested to know how well _our code that uses SQLite was tested_.

Of course, it’s very good that our SQLite-based code uses such a high-quality foundation. However, the text-based databases leverages standard C++ library and linux IO that were very well tested too.

> This testing goes way beyond the testing we do for our db files.

This is very true. Luckily text-based db does not need such a rigorous testing due to architectural simplicity.

> We have had MT issues that shown up and they are not easy to debug,... 

Please note that those MT issues affected _both_ SQLite- and text-based databases. We still need to improve tests. I was in the middle of that work when I had to postpone that for the sake of comgr and other stuff (like blocked CI).

> ...so if we can utilize something that has been well tested, it will be better and save us debugging/development time.

Yes, we wasted some time for debugging. However development time was actually saved before that: The sqlite_perfdb test and parts of the SQLite db implementation were borrowed from test-based db (together with static lock guard, unfortunately).

> We can still tweak sqlite to get better performance which will be better than inventing our own read/write caching that is not well tested.

That would be wonderful, I fully support this.

**Comment by pfultz2** [comment]
> However, the text-based databases leverages standard C++ library and linux IO that were very well tested too.

Yea but we are doing our own synchronization and caching, which is not as well tested as sqlite.

> Luckily text-based db does not need such a rigorous testing due to architectural simplicity.

It still needs rigorous testing especially when synchronizing with reads, writes, and a cache. There is no need to reimplement this when sqlite will handle these issues. 

> Please note that those MT issues affected both SQLite- and text-based databases.

More issues affect the text-based databases since it needs more synchronization. And there may still be issues, as sqlite has found cases where our locks weren't working correctly(which is why we dropped using our own locks and synchronization and just utilize sqlite). It would be good to move all of our DBs to sqlite so we can avoid possible issues with this in the future. 

> We still need to improve tests.

Yes, but we need to test where we have our custom synchronization(like `GetCached`) instead of testing reading and writing which is already taken care of by sqlite.(although I dont think these tests should be removed just to make sure we are calling into sqlite correctly).

**Comment by atamazov** [comment]
@JehandadKhan AFAICS all your review comments resolved. Anything else?

**Comment by atamazov** [comment]
@DrizztDoUrden I've upmerged this branch from develop. Please update your local branch.

**Comment by atamazov** [comment]
@DrizztDoUrden Oops, not that easy. Tidy errors emerged. Please have a look.

**Comment by atamazov** [comment]
Despite the fact that this PR has clear positive effects, it is blocked used for several months. I do not see any technical reasons for this. 

I've spent too much time and efforts for explanations that seem to have no effect.

We are discussing as if there is an SQLite-based find-db, and we propose replacing it with the "primitive" text base. However, the situation is exactly the opposite. As far as I understand, we are offered to replace the existing and fast-working implementation (from this PR) with an imaginary database that does not exist yet.

From now on, I stop discussing this topic. We can continue the discussion if someone shows me SQLite based find-db, which will be at least twice as slow as this implementation.

cc @dagamayank @daniellowell

**Comment by pfultz2** [comment]
>  I do not see any technical reasons for this.

We dont want to incorporate more MT synchronization code that has not been well tested. We have made it clear all along we want to switch to sqlite because it can improve performance and provide more correctness. Almost all dbs have transitioned over but find-db has been blocked by you.

> We are discussing as if there is an SQLite-based find-db

It is fairly trivial to change the db.

> which will be at least twice as slow as this implementation.

And there is no proof of such a claim. Even if this implementation was faster(which I am doubtful), it could be due to missing MT safety since this implementation doesn't provide the same level of testing that sqlite does.

**Comment by atamazov** [comment]
@DrizztDoUrden Please fix tidy failures.

**Comment by atamazov** [comment]
@DrizztDoUrden Build fails with OCL backend:
```
$ cmake -DCMAKE_BUILD_TYPE=Release -DMIOPEN_BACKEND=OpenCL -DMIOPEN_TEST_ALL=On -DBUILD_DEV=On
...
$ make MIOpenDriver -j 24
...
[ 29%] Building CXX object src/CMakeFiles/MIOpen.dir/readonlyramdb.cpp.o
...
.../src/include/miopen/sqlite_db.hpp:286:45: error: invalid use of incomplete type 'class miopen::SQLitePerfDb'
         -> decltype(std::declval<Derived>().FindRecordUnsafe(args...))
                     ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
...
.../src/include/miopen/sqlite_db.hpp:318:78: error: invalid use of incomplete type 'class miopen::SQLitePerfDb'
     inline auto Update(const U&... args) -> decltype(std::declval<Derived>().UpdateUnsafe(args...))
                                                      ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
```
and so on.

**Comment by atamazov** [comment]
Under rigorous testing right now.

**Comment by JehandadKhan** [comment]
> Under rigorous testing right now.

@atamazov What does that entail ? I am just curious

**Comment by atamazov** [comment]
@JehandadKhan 
> > Under rigorous testing right now.
> 
> @atamazov What does that entail ? I am just curious

In this specific case:
- Running a set of configs using MIOpenDriver, collecting logs and .csv (Pass/Fail, GPU time, Wall time, Aux Wall time, GWSS time), using verification & warm-up (`-V 1 -w 2`).
  - Baseline (develop @ da79ca10a)
    - HIP backend
      - A set of 93 FP32 configs in Find mode, MIOPEN_FIND_MODE=1
      - A set of 93 FP16 configs in Find mode, MIOPEN_FIND_MODE=1
      - A set of 93 FP32 configs in Immediate mode
      - A set of 93 FP16 configs in Immediate mode
    - OCL backend
      - A set of 93 FP32 configs in Find mode, MIOPEN_FIND_MODE=1
      - A set of 93 FP32 configs in Immediate mode
  - This PR (`ddu/cached-ufdb` @ 82817ee93)
    - The same tests with the same backends
    - Additional set of Immediate mode tests, but with find-db collected during previous step (with baseline).
- Comparing .csv and finding out how GPU time, Wall time, Aux Wall time, GWSS time are affected by the PR, for each specific config and on average.
   - Denoting progressions, regressions
   - Investigating/triaging regressions, if any (beginning with comparing the collected logs).  
   - Checking if progressions are due to expected side effects of the PR.

This is kinda "standard" set I've developed over time. It takes several hours  on my Radeon VII machine + some time for manual analysis.

For now, I am not going to repeat the full set of experiments I did for https://github.com/AMDComputeLibraries/MLOpen/pull/1968 ("best case", "worst case" etc).

**Comment by atamazov** [comment]
@DrizztDoUrden `Hip tidy` failed.

**Comment by atamazov** [comment]
@DrizztDoUrden clang tidy failure

**Comment by atamazov** [comment]
@DrizztDoUrden 
image

**Comment by atamazov** [comment]
Local tesing done. Description updated with routine "Local performance & correctness testing" and somewhat more interesting "Evaluation of Immediate Mode acceleration."

**Comment by atamazov** [comment]
@DrizztDoUrden Failed twice in a row at `Aux tests/ Hip Debug Embedded Vega20`. Please have a look.

**Comment by atamazov** [comment]
Converted to draft to avoid spoiling docker image cache on CI. You can continue work as usual. But please change this back to "normal" PR only after ending of CI moratorium (see mail). Then first merge from develop. Push the `[Ready for review]` button only after that. Thanks.

**Review by atamazov — DISMISSED** [review]

**Comment by atamazov** [comment]
@DrizztDoUrden Please resolve remaining review comments and ask @JehandadKhan to approve. We need performance. A good piece of it is here, hidden in user-find-db reads.

**Review by JehandadKhan — APPROVED** [review]
LGTM

**Comment by JehandadKhan** [comment]
@atamazov can we have someone measure how much of a difference does it make in an end-to-end network run?

**Comment by atamazov** [comment]
@Kirpich30000 Can you help with these measurements?
> @atamazov can we have someone measure how much of a difference does it make in an end-to-end network run?

**Comment by Kirpich30000** [comment]
> @Kirpich30000 Can you help with these measurements?
Should we try to ask Peng's team?

**Comment by atamazov** [comment]
@DrizztDoUrden Please fix tidy errors.

**Comment by atamazov** [comment]
> > @Kirpich30000 Can you help with these measurements?
> 
> Should we try to ask Peng's team?

Maybe. Or Kamil can help us?

/cc @sunway513 @shurale-nkn @JehandadKhan

**Comment by DrizztDoUrden** [comment]
> Vasilii Filippov Please fix tidy errors.

Sure, will do tomorrow.

**Comment by DrizztDoUrden** [comment]
Well, this didn't age well. I'll need some time to fix this.

**Comment by atamazov** [comment]
I see. Please go ahead.

File: src/include/miopen/readonlyramdb.hpp [hunk 9]
```diff
@@ -36,6 +36,21 @@
 
 namespace miopen {
 
+extern bool& testing_rordb_embed_fs_override();
+
+struct TestRordbEmbedFsOverrideLock
```
Comment (atamazov) [inline src/include/miopen/readonlyramdb.hpp:41]: Please move this definition to tests. Why: It is used in tests only.

File: src/include/miopen/readonlyramdb.hpp [hunk 10]
```diff
@@ -36,6 +36,21 @@
 
 namespace miopen {
 
+extern bool& testing_rordb_embed_fs_override();
```
Comment (atamazov) [inline src/include/miopen/readonlyramdb.hpp:39]: Please move this function into `miopen::debug` namespace. Why: It is required only for testing.

**Review by atamazov — CHANGES_REQUESTED** [review]
Thanks for fixing, but some changes still needed.

**Comment by atamazov** [comment]
Build error in tidy

**Comment by atamazov** [comment]
@DrizztDoUrden Ping

**Comment by DrizztDoUrden** [comment]
Merged.

**Comment by atamazov** [comment]
@DrizztDoUrden Thanks for this great work! Now I will test it and pass to @changpeng for testing with some NAS-like frameworks.

**Comment by atamazov** [comment]
## :cyclone: Testing in progress...

**Comment by atamazov** [comment]
## :cyclone: Testing results

gfx906/60, ROCm 4.0, 93 FP32 configs, all directions, performance gain:

&nbsp; | GPU time | Wall Time | Aux Wall time | GWSS time(1)
-- | -- | -- | -- | --
Normal Find, user fdb empty, binary cache empty | 1% | 16% | 11% | 345%
Fast Find(2), user fdb full, binary cache full | 0% | 7% | 49% | 69%

- (1) Time spent for `GetWorkspaceSize()`
- (2) User find-db is full, so it re-uses find-db records

**Review by atamazov — APPROVED** [review]
It's good to go, but let's wait for feedback from @sunway513

File: src/ramdb.cpp [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+    MIOPEN_LOG_I2("Trying to store record at key " << key << " in cache for file "
+                                                   << GetFileName());
+    const auto lock = exclusive_lock(GetLockFile(), GetLockTimeout());
+    MIOPEN_VALIDATE_LOCK(lock);
+
+    if(!DisableUserDbFileIO)
+    {
+        if(!StoreRecordUnsafe(record))
+            return false;
+        UpdateDbModificationTime(GetFileName());
```
Comment (zpwenjh) [inline src/ramdb.cpp:150]: This `UpdateDbModificationTime` function will update the modification time of user_find_db. Thus `time` > `file_read_time` in the following function `UpdateCacheEntryUnsafe`, which result in `is_valid` being false, then, the code block in the if statement is not executed and `file_read_time` not updated. The next time finding needs to prefetch() again. 
```
const auto is_valid = ValidateUnsafe();

    if(!DisableUserDbFileIO)
        UpdateDbModificationTime(GetFileName());

    if(is_valid)
    {
        const auto& key = record.GetKey();
        const auto it   = cache.find(key);
        auto ss         = std::ostringstream{};
        record.WriteIdsAndValues(ss);

        if(it != cache.end())
        {
            auto& item   = it->second;
            item.content = ss.str();
        }
        else
        {
            cache.emplace(key, CacheItem{-1, ss.str()});
        }
        file_read_time = ramdb_clock::now();
    }
```
But what you want is not to prefetch again if there is no other thread to modify the `user_find_db`, right?

**Comment by zpwenjh** [comment]
> Extra feature: the file I/O can be switched off by the build flag (MIOPEN_DISABLE_USERDB). This is intended for the systems where we do not want to read/write files. In this mode, **the cache works as a non-persistent user-find-db**, thus partially enabling functionalities that leverage user-find-db. The most notable one is Hybrid Find mode.

The db object is created here:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/2b9f0ea8f74ac24a215ad73395b41c45b920cb49/src/include/miopen/find_db.hpp#L86-L120

The path of user_find_db is from `GetUserPath()`:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/2b9f0ea8f74ac24a215ad73395b41c45b920cb49/src/find_db.cpp#L210-L220

If `MIOPEN_DISABLE_USERDB` is off, then the path of user_find_db is empty. Thus, **the cache works as a non-persistent user-find-db** is untenable. Is it necessary to correct the code in order to match this? I also think cache works as a non-persistent user-find-db is very helpful.

File: src/ramdb.cpp (hunk 11 above)
Comment (junliume) [reply src/ramdb.cpp:150]: @atamazov and @DrizztDoUrden , could you take a look at the above comments?

**Comment by junliume** [comment]
> > Extra feature: the file I/O can be switched off by the build flag (MIOPEN_DISABLE_USERDB). This is intended for the systems where we do not want to read/write files. In this mode, **the cache works as a non-persistent user-find-db**, thus partially enabling functionalities that leverage user-find-db. The most notable one is Hybrid Find mode.
> 
> The db object is created here:
> 
> https://github.com/ROCmSoftwarePlatform/MIOpen/blob/2b9f0ea8f74ac24a215ad73395b41c45b920cb49/src/include/miopen/find_db.hpp#L86-L120
> 
> If `MIOPEN_DISABLE_USERDB` is off, in the immediate mode, `MultiFileDb` will not create `user db` and in the find modes, the `user db` will not create too. So there is no cache for the following find result. Thus, **the cache works as a non-persistent user-find-db** is untenable. Is it necessary to correct the code in order to match this? I also think cache works as a non-persistent user-find-db is very useful to avoid calling `TryLoad()` funtion again and again while `MIOPEN_DISABLE_USERDB` is off.

@zpwenjh could you please create an issue for better visibility? e.g. [Post Merge Issue] ram DB issue from #280

**Comment by atamazov** [comment]
@zpwenjh We see this, no problem. Need to find some free time to answer.

File: src/include/miopen/temp_file.hpp [hunk 12]
```diff
@@ -38,8 +38,18 @@ class TempFile
...
 
-    inline std::string Path() const { return (dir.path / name).string(); }
-    inline operator std::string() const { return Path(); }
+    TempFile(TempFile&& other) noexcept : name(std::move(other.name)), dir(std::move(other.dir)) {}
+
+    TempFile& operator=(TempFile&& other) noexcept
+    {
+        name = std::move(other.name);
+        dir  = std::move(other.dir);
+        return *this;
```
Comment (atamazov) [inline src/include/miopen/temp_file.hpp:47]: :anchor:
